### DISCUSSION TITLE
Crop or Trim images

### DISCUSSION THREAD

**Discussion by TNNSLAB** [ImageMagick/ImageMagick#3833] [Help] [answered]
So I got a lot of fashion images like the ones attached, and I have to crop them/trim them in photoshop to 3:4 ratio. 
I want to automate this process and wondering if Imagemagick could help me with that. 
Center the model and crop it/trim it to 3:4 ratio.

![crop-1](https://user-images.githubusercontent.com/46636486/122206501-ba902f00-cea1-11eb-86b8-475d94530e47.jpg)
![crop-2](https://user-images.githubusercontent.com/46636486/122206519-c0861000-cea1-11eb-8063-514ef54a66fd.jpg)
![crop-3](https://user-images.githubusercontent.com/46636486/122206544-ca0f7800-cea1-11eb-8916-fae3376dd0d9.jpg)
![crop-4](https://user-images.githubusercontent.com/46636486/122206563-d1368600-cea1-11eb-9244-e5bb3b87ca86.jpg)

**Comment by snibgo** [comment]
IM can do whatever you want. For example, trimming the final image:
```
magick 122206563-d1368600-cea1-11eb-9244-e5bb3b87ca86.jpg -fuzz 10% -trim +repage x.png
```
The result is 1390x3073 pixels.

> Center the model and crop it/trim it to 3:4 ratio.

This seems contradictory. Please show the required result for the last image.

**Reply by TNNSLAB** [reply]
@snibgo thank you for you quick response ! glad to know that IM can do everything ! 

I forgot to mention about the background colour. It needs to be be the same. Normally i think it will stay around 240-255 RGB. 
Final image should look like this : 
![cropped_to_3-4](https://user-images.githubusercontent.com/46636486/122240644-38643280-cec2-11eb-84c7-91922ed999e9.png)

**Comment by fmw42** [comment]
I believe you can do that in ImageMagick 7 as follows in Unix syntax:

```
magick 122240644-38643280-cec2-11eb-84c7-91922ed999e9.png \
-set option:color "%[pixel:u.p{0,0}]" \
-fuzz 5% -trim +repage \
-set option:dims "%[fx:h/w>4/3?h*3/4:w]x%[fx:h/w>4/3?h:w*4/3]" \
-gravity center -background "%[color]" \
-extent "%[dims]" \
result.png

```

![result](https://user-images.githubusercontent.com/13458218/122245889-f663ce00-ce7a-11eb-8527-3141c2061ed5.png)

**Reply by TNNSLAB** [reply]
@fmw42  getting close, but the background is not right : 

![result](https://user-images.githubusercontent.com/46636486/122246538-ea9df900-cec6-11eb-9f10-213daa509522.png)

**Reply by TNNSLAB** [reply]
Version: ImageMagick 7.0.10-25 Q16 x86_64 2020-08-01 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib

**Reply by fmw42** [reply]
Worked fine for me.  What is your platform/OS? What was your exact command line? Is this the same image that I used?

**Reply by TNNSLAB** [reply]
@fmw42 any ideas why ? 
![result](https://user-images.githubusercontent.com/46636486/122248442-5b91e080-cec8-11eb-8cd1-763ca009ebf1.png)

**Reply by TNNSLAB** [reply]
Mac Os Mojave 10 :(  I think u are running the wrong file as input...try one of the very first 3 images please...

**Reply by TNNSLAB** [reply]
yup, running the same command exactly

**Reply by fmw42** [reply]
First image:

magick 122206501-ba902f00-cea1-11eb-86b8-475d94530e47.jpg \
-set option:color "%[pixel:u.p{0,0}]" \
-fuzz 5% -trim +repage \
-set option:dims "%[fx:h/w>4/3?h*3/4:w]x%[fx:h/w>4/3?h:w*4/3]" \
-gravity center -background "%[color]" \
-extent "%[dims]" \
result.png


![result](https://user-images.githubusercontent.com/13458218/122249779-fb764c80-ce7d-11eb-9e65-6cfbe39de7d7.png)

Looks fine to me. I am on Mac OSX Sierra running IM 7.1.0.0 Q16

**Reply by fmw42** [reply]
Please post your exact code that you used and your result for this same image. Perhaps you have a typo.

**Reply by fmw42** [reply]
Your IM version 7 is very old. Perhaps upgrade and try again, if not working properly.

**Reply by TNNSLAB** [reply]
@fmw42 very strange behavior. Working great on 122206501-ba902f00-cea1-11eb-86b8-475d94530e47.jpg  but fails on this : 122206544-ca0f7800-cea1-11eb-8916-fae3376dd0d9.jpg 

you think this might be because of the position of the model more to the left ?

**Reply by fmw42** [reply]
I do not know which image that is. Please post it and your result and exact command line so I can see if there is a typo.

**Reply by TNNSLAB** [reply]
@fmw42 ok, I tried it with this image : 

![test4](https://user-images.githubusercontent.com/46636486/122380004-b595b300-cf67-11eb-8bca-3fb391235524.jpg)
and I'm getting this : 
![result](https://user-images.githubusercontent.com/46636486/122380041-c0504800-cf67-11eb-961f-ba8ac9c9eb25.png)

This is the code i run. Is there a way to have the background even ? without those lines ? 

`magick test4.jpg \
-set option:color "%[pixel:u.p{0,0}]" \
-fuzz 5% -trim +repage \
-set option:dims "%[fx:h/w>4/3?h*3/4:w]x%[fx:h/w>4/3?h:w*4/3]" \
-gravity center -background "%[color]" \
-extent "%[dims]" \
result.png`

**Comment by snibgo** [comment, marked as answer]
"-extent" uses a single colour for the added boundary. Instead, we can use a viewport and virtual-pixel edge, so the added boundary takes the colour from the edge of the image, so we don't get a discontinuity.

Viewports are not sensitive to "-gravity", so to centre the image we need to set offsets.

Windows BAT syntax, so adjust for bash:
```
magick ^
  model4.jpg ^
  -fuzz 5%% -trim +repage ^
  -set option:W1 "%%[fx:h/w>4/3?h*3/4:w]" ^
  -set option:H1 "%%[fx:h/w>4/3?h:w*4/3]" ^
  -set option:X1 "%%[fx:((h/w>4/3?h*3/4:w)-w)/2]" ^
  -set option:Y1 "%%[fx:((h/w>4/3?h:w*4/3)-h)/2]" ^
  -set option:distort:viewport %%[W1]x%%[H1]-%%[X1]-%%[Y1] ^
  -virtual-pixel edge ^
  -filter point -distort srt 1,0 ^
  +repage ^
  x3.png
```

**Reply by TNNSLAB** [reply]
for some reason the result is not 3:4 
![x3](https://user-images.githubusercontent.com/46636486/122393430-28595b00-cf75-11eb-8db8-c12a5d9597c8.png)

**Comment by snibgo** [comment]
For me, the size of the result is 4752x6336, which is exactly 3:4. Your result is 3604x6336, which is the size I get for the trimmed image, before viewport. So I suspect a bug in your command at that stage. 

There may be a mistake in the translation from Windows BAT to bash. Please post the bash script you used.

**Reply by TNNSLAB** [reply]
`magick \
  model.jpg \
  -fuzz 5%% -trim +repage \
  -set option:W1 "%%[fx:h/w>4/3?h*3/4:w]" \
  -set option:H1 "%%[fx:h/w>4/3?h:w*4/3]" \
  -set option:X1 "%%[fx:((h/w>4/3?h*3/4:w)-w)/2]" \
  -set option:Y1 "%%[fx:((h/w>4/3?h:w*4/3)-h)/2]" \
  -set option:distort:viewport %%[W1]x%%[H1]-%%[X1]-%%[Y1] \
  -virtual-pixel edge \
  -filter point -distort srt 1,0 \
  +repage \
  x3.png`

**Reply by TNNSLAB** [reply]
how i can format the code properly here :(

**Comment by snibgo** [comment]
In Windows BAT scripts, perent % is a special character, so needs to be doubled %%.

In bash, percent % is not a special character, so don't double it.

**Reply by TNNSLAB** [reply]
Fantastic ! Can't thank you enough !
